MinecraftClient/ChatBots/AutoCarft.cs

```diff
@@ -0,0 +1,697 @@
+                    waitingForMaterials = false;
+                    craftingFailed = false;
+                }
+                HandleNextStep();
+            }
+        }
+
+        private void OpenTable(Location location)
+        {
+            SendPlaceBlock(location, Direction.Up);
+        }
+
+        /// <summary>
+        /// Prepare the crafting action steps by the given recipe name and start crafting
+        /// </summary>
+        /// <param name="recipe">Name of the recipe to craft</param>
+        private void PrepareCrafting(string name)
+        {
+            PrepareCrafting(recipes[name]);
+        }
+
+        /// <summary>
+        /// Prepare the crafting action steps by the given recipe and start crafting
+        /// </summary>
+        /// <param name="recipe">Recipe to craft</param>
+        private void PrepareCrafting(Recipe recipe)
+        {
+            recipeInUse = recipe;
+            if (recipeInUse.CraftingAreaType == ContainerType.PlayerInventory)
+                inventoryInUse = 0;
+            else
+            {
+                var inventories = GetInventories();
+                foreach (var inventory in inventories)
+                    if (inventory.Value.Type == ContainerType.Crafting)
+                        inventoryInUse = inventory.Key;
+                if (inventoryInUse == -2)
+                {
+                    // table required but not found. Try to open one
+                    OpenTable(tableLocation);
+                    waitingForTable = true;
+                    SetTimeout("table not found");
+                    return;
+                }
+            }
+
+            foreach (KeyValuePair<int, ItemType> slot in recipe.Materials)
+            {
+                // Steps for moving items from inventory to crafting area
+                actionSteps.Add(new ActionStep(ActionType.LeftClick, inventoryInUse, slot.Value));
+                actionSteps.Add(new ActionStep(ActionType.LeftClick, inventoryInUse, slot.Key));
+            }
+            if (actionSteps.Count > 0)
+            {
+                // Wait for server to send us the crafting result
+                actionSteps.Add(new ActionStep(ActionType.WaitForUpdate, inventoryInUse, 0));
+                // Put item back to inventory. (Using shift-click can take all item at once)
+                actionSteps.Add(new ActionStep(ActionType.ShiftClick, inventoryInUse, 0));
+                // We need to wait for server to update us after taking item from crafting result
+                actionSteps.Add(new ActionStep(ActionType.WaitForUpdate, inventoryInUse));
+                // Repeat the whole process again
+                actionSteps.Add(new ActionStep(ActionType.Repeat));
+                // Start crafting
+                ConsoleIO.WriteLogLine("AutoCraft start!");
+                HandleNextStep();
+            }
+            else ConsoleIO.WriteLogLine("AutoCraft cannot be started. Check your available materials");
```

Comment:
`"AutoCraft cannot be started. Check your available materials for crafting " + recipe.ResultItem`